src/psygnal/_group.py

```diff
@@ -237,17 +190,170 @@ def disconnect(self, slot: Callable | None = None, missing_ok: bool = True) -> N
+# a container for user-determined names.  If names must be added, try to prefix
+# with "psygnal_" to avoid conflicts with user-defined names.
+@mypyc_attr(allow_interpreted_subclasses=True)
+class SignalGroup:
+    """A collection of signals that can be connected to as a single unit.
+
+    This class is not intended to be instantiated directly.  Instead, it should be
+    subclassed, and the subclass should define Signal instances as class attributes.
+    The SignalGroup will then automatically collect these signals and provide a
+    SignalRelay instance that can be used to connect to all of the signals in the group.
+
+    This class is used in both the EventedModels and the evented dataclass patterns.
+    See also: `psygnal.SignalGroupDescriptor`, which provides convenient and explicit
+    way to create a SignalGroup on a dataclass-like class.
+
+    Parameters
+    ----------
+    instance : Any, optional
+        An object to which this `SignalGroup` is bound, by default None
+
+    Attributes
+    ----------
+    all : SignalRelay
+        A special SignalRelay instance that can be used to connect to all signals in
+        this group.  The name of this attribute can be overridden by the user by
+        creating a new name for the SignalRelay annotation on a subclass of SignalGroup
+        e.g. `my_name: SignalRelay`
+
+    Examples
+    --------
+    ```python
+    from psygnal import Signal, SignalGroup
+
+    class MySignals(SignalGroup):
+        sig1 = Signal()
+        sig2 = Signal()
+
+    group = MySignals()
+    group.all.connect(print) # connect to all signals in the group
+
+    list(group)                  # ['sig1', 'sig2']
+    len(group)                   # 2
+    group.sig1 is group['sig1']  # True
+    ```
+    """
+
+    _psygnal_signals: ClassVar[Mapping[str, Signal]]
+    _psygnal_instances: dict[str, SignalInstance]
+    _psygnal_uniform: ClassVar[bool] = False
+
+    # see comment in __init__.  This type annotation can be overriden by subclass
+    # to change the public name of the SignalRelay attribute
+    all: SignalRelay
+
+    def __init__(self, instance: Any = None) -> None:
+        cls = type(self)
+        if not hasattr(cls, "_psygnal_signals"):  # pragma: no cover
+            raise TypeError(
+                "Cannot instantiate SignalGroup directly.  Use a subclass instead."
+            )
+        self._psygnal_instances: dict[str, SignalInstance] = {
+            name: signal.__get__(self, cls)
+            for name, signal in cls._psygnal_signals.items()
+        }
+        self._psygnal_relay = SignalRelay(self._psygnal_instances, instance)
+
+        # determine the public name of the signal relay.
+        # by default, this is "all", but it can be overridden by the user by creating
+        # a new name for the SignalRelay annotation on a subclass of SignalGroup
+        # e.g. `my_name: SignalRelay`
+        self._psygnal_relay_name = "all"
```

Comment:
How can we define a custom name from `SignalGroupDescriptor` ?
A keyword argument to `__init__` ? Allow parsing instanciated `SignalRelay` attributes, so we can pass it with `type("MyGroup", (SignalGroup, ), {"relay": SignalRelay()})` ?

Reply:
as described in the comment above and in the class docstring:  as a SignalRelay annotation in the class:

```python
class MyGroup(SignalGroup):
    custom_relay_name: SignalRelay 

# or, if you want to do it dynamically, which I'm less concerned about syntax
type("MyGroup", (SignalGroup, ), {"__annotations__": {'custom_relay_name': SignalRelay} })
```

The reason for doing it this way is that it will satisfy type checkers as well.  If we take an argument in `__init__`, then the end-user won't get any type-hinting or autocompletion for that name. If they use the syntax above, they  will get type checking and completion.

Reply:
(linking to https://github.com/pyapp-kit/psygnal/pull/269#issuecomment-1947225024 ... it's possible we don't want this mechanism at all if we're going to provide a permanent public name)

Reply:
```python
type("MyGroup", (SignalGroup, ), {"__annotations__": {'custom_relay_name': SignalRelay}, "name": Signal(str, str) })
```
Then `MyGroup.name` does not have type annotations, we would also need to add the annotation to `"__annotations__"`.

Anyway, if needed, all this boilerplate code can be put in a class method.

Reply:
> Then MyGroup.name does not have type annotations

it *does* have annotations as far as psygnal is concerned... which is simple to look at `cls.__annotations__` in [this one line here](https://github.com/pyapp-kit/psygnal/pull/269/files#diff-8218e71615e3cb1b30f78bd2143ca77661925d39178dd9ccad62dc0f62206c98R271).  But if you mean that mypy and an IDE wouldn't give you annotations, yes that's true, but any dynamic type creation is already hopeless for those things.  
so the point is:  it's always best practice for a user to hint this with a type annotation, and even if we did offer a parameter in `__init__`, they might *still* add that type annotation just for the sake of mypy.  So, we might as well use that mechanism anyway (if we're going to do this at all that is)

Reply:
```python
In [1]: class MyGroup(SignalGroup):
   ...:     custom_name: SignalRelay
   ...:

In [2]: Group2 = type("Group2", (SignalGroup, ), {"__annotations__": {'custom_relay_name': SignalRelay}, "name": Signal(str, str) })

In [3]: MyGroup.__annotations__
Out[3]: {'custom_name': <class 'psygnal._group.SignalRelay'>}

In [4]: Group2.__annotations__
Out[4]: {'custom_relay_name': <class 'psygnal._group.SignalRelay'>}

In [5]: MyGroup().custom_name
Out[5]: <SignalRelay>

In [6]: Group2().custom_relay_name
Out[6]: <SignalRelay>
```

Reply:
> Then MyGroup.name does not have type annotations

oops sorry. I think I misread this the first time.  You're right that 'name' would also need to go in the `__annotations__` dict... but how many dynamically created classes are that worried about type annotations?